### ISSUE TITLE
Proposal for a new incompatible algorithm

### ISSUE THREAD

**Issue by 0x2b3bfa0** [lesspass/lesspass#389]
I would like to write down some ideas for a possible future algorithm, improving some parts of the existing one.

* Hashing of the master password along with a master password counter value and then hashing the result along with the site-specific information: this would allow to store the master password in hardware keys and trusted platform modules without disclosing the plaintext. If the hash were compromised, the user only would need to increment the master password counter. This would allow to bring some optional security improvements in a near future without breaking the algorithm again. See #376, #145 and #387 for examples.

* Hashing one time each value before concatenating them, in order to fix #336 and prevent (hypothetical) bugs related to character encoding and JSON parsing or escaping issues.

* Enforce the use of UTF-8 when transforming the strings entered by the user into raw octet data. I don't know how this potential inconsistency is being handled, nor whether it can lead to different output between platforms or locale settings.

I think that is necessary to discuss deeply the convenience of changing again the algorithm, and have months (or perhaps years) of technical and practical thinking to avoid having to change a time more in the next century.

Feel free to close this issue if is (currently) out of scope.

**Comment by guillaumevincent** [comment]
We can open the discussion here about the version 3 for LessPass.

We should speak about replacing pbkdf2 by argon2. And we should talk about the behavior for the site field (autocompletion and autofill)

For hardware key, maybe we can instead store a hash used in combination with the master password. It will allow us to add the key in configuration for the webextension for example

**Comment by 0x2b3bfa0** [comment]
Note to self: replacing PBKDF2 by another function is already discussed on #56, but we should take in account the browser support for heavy libraries and the use of polyfills.

**Comment by claudiusraphaelpaeth** [comment]
> * store the master password in hardware keys and trusted platform modules without disclosing the plaintext. If the hash were compromised, the user only would need to increment the master password counter

Was my thought since i stumbled over the first version of lesspass. Most definitely the one thing that would make the foundation feature-complete.

Allowing to completely separate the Master-password/key from the generator and allows to use the latter in secondary external devices, which could be fed optically via e.g.: qr shown by the add-on and also vice versa be transferred for general usage via virtualized human-interface-device, for example keyboard-emulation via bluetooth.

From there on it would allowing to decide for a custom/manual set and load of function of choice and to allow to use a network-independent device as well, like an old smartphone or a smartwatch for example, which most often need to take other considerations into account then the mass of used devices which shall be served by the by default version of lesspass, which shall be when feasible updated according to the possibilities of the generations in and coming to use; without breaking compatibility regarding hardware-dependency for virtualized HID and/or use with old devices which might not offer the necessary feature-set and/or power for future functions.

Reaching this stage and basic definition would be the pinnacle of what is possible for the concept "lesspass" and full closure in terms of the ideal and resulting differences against nearly all other existing password-vault/manager-concepts. At that point also possible to be fully decoupled and modularized, distilled into equal core components, that can be freely and in a controlled manner combined to a fixed set of combinations with known limitations, not allowing to introduce bugs into the core-park.

Only a thought.

Just wanted to give a thumbs-up, especially for the first part of the proposal "decoupling masterkey from generator completely", as well as considering to update function and therefore make use of the abilities of current and future devices.

**Comment by claudiusraphaelpaeth** [comment]
An addition regarding the decoupling of the Master-Keyword:

+ For now it would be considerable to start with a separate application for use on PC which can then finally be ported to an android/etc. app.

- Its purpose would be solely to let a user input the Master-Keyword
and

- to calculate a pad of disposable one-time keys
as well as

- if it is possible to check from the add-ons in which exact browser they do run, to mark the pad of keys for usage in specific browsers as a matter of enhanced security

… to minimize attack-surface, but also to take the interdependency of the specific browsers into question; Chrome syncs, Chromium doesn't sync necessarily (depending on the build), also Chrome offers to generate keys/passwords and store them in the Google-Profile, which are by default encrypted against the Google-Servers unless one chooses to use a passphrase, against which the key-vault will be encrypted then. I don't know about FireFox's  actual behaviour, or how other browsers handle it. But this shows that it could be critical to diverse the generated pads to target specific browsers.

Please let me know if you decide to start working on a comparable solution, I would like to be part of it.

Thanks for reading. Hope it helps to find a decision or initiating new thoughts.

**Comment by Charon77** [comment]
I'd like to address the behaviour of autocompletion and autofill. I think this is a really great enhancement I would want to use, mostly on mobile phones like Android. Yes, I am aware that there are attacks that targets these kind of usage as it injects to the webpage, but I believe that users still benefit from this. A paranoid user may opt out from this feature (or conversely, make this a normally-inactive feature, and warn user about potential security breach before activing this.) Clipboard may also be addressed in similar matter.

**Comment by Charon77** [comment]
Argon2 seems to be a nice choice for the next algorithm. It's flexible on how much memory and computation time should be used by the algorithm, so I think this allows anywhere from browser implementation to beefy CPUs to calculate the password hash.

I think users should be allowed to chose on the derivation strength being used. However, this would also mean that users will have to know that the these strength are not interchangable.

**Comment by guillaumevincent** [comment]
I don't know how argon has a good support on web platform. To increase the security, we should add a generated key to add some entropy to the master password.

**Comment by edouard-lopez** [comment]
* #376 talks about 2FA
* https://github.com/lesspass/lesspass/issues/56#issuecomment-627873344 `lib-sodium` support `Argon2` (among other)

**Comment by guillaumevincent** [comment]
We could also look at wasm component from rust lib

**Comment by elgow** [comment]
For version 3 can you please provide the exclude character feature in the javascript versions as well as the CLI. My greatest frustration with lesspass remains having to constantly battle with sites that don't support the same list of special characters that lesspass has chosen. 

I seem to be running at about 50% of sites refusing lesspass generated passwords. The usual result is that I have to save the settings for the site in the DB to remind me to disable specials and then remember to add a single special to the generated password in order to satisfy site password policy. This is a lot of trouble, and results in a weaker password because the special chars have been turned of in password generation. Please correct this.

**Comment by guillaumevincent** [comment]
Yes and no. Yes because the rule/profile associated with a site should add the maximum of available characters. But no because it's not part of the algorithm part. It's more a UI/UX issue. With the v2 we can already imagine having a way to choose a subset of special chars.

**Comment by elgow** [comment]
Thanks @guillaumevincent. I agree this could be solved in V2 by adding a field for chars to be excluded from the generated password as you've done in the CLI. That field would need to be stored in the DB to be really useful, but can default to nothing and thus not be a breaking change.

**Comment by lolo32** [comment]
More and More website request to use two factor authentication (2fa) so it would great to store the secret in an encrypted form and that LessPass can provide it.

**Comment by guillaumevincent** [comment]
@lolo32 I'm not sure I understand. There is a sync mechanism in 2FA. Secret is different every time you generate it.
